File: Easy/isHappy.java
// Leetcode-202
// Happy Number
// https://leetcode.com/problems/happy-number/description/


class Solution {
  public static boolean isHappy(int n) {
        int fast = n;
        int slow = n;
        do{
            slow = getSquareOfDigit(slow);
            fast = getSquareOfDigit(getSquareOfDigit(fast));
        }while(fast != slow);
        if(fast ==1)
        return true;
        return false;
    }

    public static int getSquareOfDigit(int n){
        int sum =0;
        while(n > 0){
            int digit = n %10;
            sum = sum + digit * digit;
            n = n/10;
        }
        return sum;
    }
}

File: Easy/isHappy.js
/**
 * Leetcode-202
 * Happy Number
 * https://leetcode.com/problems/happy-number/
 * #Two pointers
 */

var isHappy = function (n) {
  slow = n;
  fast = n;

  do {
    slow = helper(slow);
    fast = helper(helper(fast));
    if (fast == 1) {
      return true;
    }
  } while (slow != fast);

  return false;
};

var helper = function (num) {
  let sum = 0;
  while (num != 0) {
    let rem = num % 10;
    sum = sum + rem * rem;
    num = Math.floor(num / 10);
  }
  return sum;
};

console.log(isHappy(19));
console.log(isHappy(2));
